distributed/shuffle/_rechunk.py

```diff
@@ -441,47 +447,98 @@ def _() -> dict[str, tuple[NDIndex, bytes]]:
         self.raise_if_closed()
-        assert meta is None
-        assert self.transferred, "`get_output_partition` called before barrier task"
+        if not self.transferred:
+            raise RuntimeError("`get_output_partition` called before barrier task")
 
         await self._ensure_output_worker(partition_id, key)
-
         await self.flush_receive()
-
         data = self._read_from_disk(partition_id)
-
-        def _() -> np.ndarray:
-            return convert_chunk(data)
-
-        return await self.offload(_)
+        return await self.offload(convert_chunk, data)
 
     def _get_assigned_worker(self, id: NDIndex) -> str:
         return self.worker_for[id]
 
 
+@dataclass
+class ArrayRechunkSpec(ShuffleSpec):
+    new: ChunkedAxes
+    old: ChunkedAxes
+
+
+@dataclass
+class ArrayRechunkRunSpec(ShuffleRunSpec):
+    new: ChunkedAxes
+    old: ChunkedAxes
+
+
 @dataclass(eq=False)
-class ArrayRechunkState(ShuffleState):
-    type: ClassVar[ShuffleType] = ShuffleType.ARRAY_RECHUNK
+class ArrayRechunkState(SchedulerShuffleState):
     worker_for: dict[NDIndex, str]
     old: ChunkedAxes
     new: ChunkedAxes
 
-    def to_msg(self) -> dict[str, Any]:
-        return {
-            "status": "OK",
-            "type": ArrayRechunkState.type,
-            "run_id": self.run_id,
-            "worker_for": self.worker_for,
-            "old": self.old,
-            "new": self.new,
-            "output_workers": self.output_workers,
-        }
-
 
-def get_worker_for_hash_sharding(
+@spec_to_scheduler_state.register(ArrayRechunkSpec)  # type: ignore[misc]
+def _array_rechunk_from_spec(
+    spec: ArrayRechunkSpec, plugin: ShuffleSchedulerPlugin
+) -> ArrayRechunkState:
+    parts_out = product(*(range(len(c)) for c in spec.new))
+    mapping = plugin._pin_output_workers(
+        spec.id, parts_out, _get_worker_for_hash_sharding
+    )
+    output_workers = set(mapping.values())
+
+    return ArrayRechunkState(
+        id=spec.id,
+        run_id=next(SchedulerShuffleState._run_id_iterator),
+        worker_for=mapping,
+        output_workers=output_workers,
+        old=spec.old,
+        new=spec.new,
+        participating_workers=output_workers.copy(),
+    )
+
+
+@scheduler_state_to_run_spec.register(ArrayRechunkState)  # type: ignore[misc]
+def _array_state_to_run_spec(state: ArrayRechunkState) -> ArrayRechunkRunSpec:
+    return ArrayRechunkRunSpec(
+        id=state.id,
+        run_id=state.run_id,
+        worker_for=state.worker_for,
+        output_workers=state.output_workers,
+        new=state.new,
+        old=state.old,
+    )
+
+
+@run_spec_to_worker_run.register(ArrayRechunkRunSpec)  # type: ignore[misc]
+def _array_run_spec_to_run(
+    spec: ArrayRechunkRunSpec, plugin: ShuffleWorkerPlugin
+) -> ShuffleRun:
+    return ArrayRechunkRun(
+        worker_for=spec.worker_for,
+        output_workers=spec.output_workers,
+        old=spec.old,
+        new=spec.new,
+        id=spec.id,
+        run_id=spec.run_id,
```

Comment:
It seems sad that the `ArrayRechunkRun` object loses all the grouping of information (e.g. why do we not just pass the `spec` object in here?)

Reply:
I'm currently working on a slightly alternative approach that tries to roundtrip the initial `ShuffleSpec` and avoid some subclassing in the process.